### PR TITLE
Camera-lidar calibration tools initial release

### PR DESCRIPTION
**Pull request by knzo25**
## PR Type

- Improvement

## Related Links

 - Part of the release of the camera-lidar calibration tools: [PR](https://github.com/tier4/calibration_tools/pull/116)
 - Compatible with the new data from Kashiwa: [Ticket](https://tier4.atlassian.net/browse/T4PB-14331)
 - Adjusted parameters: [Ticket](https://tier4.atlassian.net/browse/T4PB-14153)

## Description

This PR implements several improvements upon lidartag to improve its detection capabilities.
It now provides very high detection rates and high accuracy poses for tags places up to 12 meters, and close tags with up with about half of the tag outside the dense zone.
* I have not tested yet the numerical performance when the tags face away from the lidar

New features:
 - Fix clusters that were broken due to the one-pass linking strategy.
 - Replaced the 4x line RANSAC algorithm since finding corners with this algorithm provided several limitations by a new rectangle estimation model (QR decomposition and SVD-based algorithm) with an optional ad-hoc RANSAC.
 - Fixed the initial edge group classification used during corner estimation using the rectangle model with the RANSAC hypotheses. It provides near-perfect edge classification (PCA is supposed to find the correct axes, but in practice does not provide good-enough results).
 - Final camera tools compliance.

## Review Procedure

 - Use both GLP and Kashiwa data and confirm that the lidartag detection rate is high (~95% or higher)
 - Launch lidartag using the calibration tools launchers
 - Confirm the calibration pairs: front_lower\camera6, right_upper\camera0, and right_upper\camera1

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Code follows [coding guidelines][coding-guidelines]
- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code follows [coding guidelines][coding-guidelines]
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets
- [ ] Write [release notes][release-notes]
- [ ] (When added something to `.repos`) Check if proper access rights are set

## CI Checks

- **vcs import**: Required to pass before the merge.
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[coding-guidelines]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/1194394777/T4
[release-notes]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/563774416
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

**Comment by yabuta** [comment]
@sgk-000 @aohsato 時間がある時にレビューをお願いします

**Review by sgk-000 — APPROVED** [review]

**Comment by sgk-000** [comment]
すいません、間違えてapproveしてしまいました。今レビュー中です

**Comment by sgk-000** [comment]
現時点だとtag size=0.8,1.03の２つのtagを使うのがデフォルトになっていますが、この点は検討が必要かもしれません。
しばらくは0.8しか使わないのであれば、tag sizeは0.8のみで良いかもしれません。

**Comment by knzo25** [comment]
@sgk-000 
次のPRで既にそう設定しました

**Review by sgk-000 — APPROVED** [review]
LGTM
柏の葉、GLPのデータで動作確認済み

**Comment by knzo25** [comment]
このPRで @sgk-000  のreviewで大丈夫ですか？
@yabuta @aohsato

File: src/lidartag.cpp [hunk 1]
```diff
@@ -2149,6 +2246,144 @@ Eigen::Vector3f LidarTag::estimateEdgeVector(ClusterFamily_t & cluster)
...
+
+    cloud1->height = 1;
+    cloud2->height = 1;
+    cloud3->height = 1;
+    cloud4->height = 1;
+
+    //pcl::io::savePCDFileASCII ("cloud1.pcd", *cloud1);
+    //pcl::io::savePCDFileASCII ("cloud2.pcd", *cloud2);
+    //pcl::io::savePCDFileASCII ("cloud3.pcd", *cloud3);
+    //pcl::io::savePCDFileASCII ("cloud4.pcd", *cloud4);
```
Comment (aohsato) [inline src/lidartag.cpp]: 不要？

File: include/lidartag/lidartag.hpp [hunk 2]
```diff
@@ -486,6 +494,25 @@ class LidarTag: public rclcpp::Node {
    */
   void organizeDataPoints(ClusterFamily_t & t_cluster);
 
+  /* <A cluster>
+   * A function to estimate the four corners of a tag using 4x line RANSAC
+   */
+  bool estimateCornersUsingLinesRANSAC(ClusterFamily_t & t_cluster,
+    pcl::PointCloud<pcl::PointXYZ>::Ptr & cloud1, pcl::PointCloud<pcl::PointXYZ>::Ptr & cloud2,
+    pcl::PointCloud<pcl::PointXYZ>::Ptr & cloud3, pcl::PointCloud<pcl::PointXYZ>::Ptr & cloud4,
```
Comment (aohsato) [inline include/lidartag/lidartag.hpp:502]: 1~4 より良い名前はない？

File: scripts/rectangle_model.py [hunk 3]
```diff
@@ -0,0 +1,423 @@
+from cProfile import label
```
Comment (aohsato) [inline scripts/rectangle_model.py]: アルゴリズム検証用のコード？ここにいらないかも？

File: src/rectangle_estimator.cpp [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <lidartag/rectangle_estimator.hpp>
+#include <algorithm>
+// /#include <ranges>
```
Comment (aohsato) [inline src/rectangle_estimator.cpp]: 不要なコメント削除 → ファイル全体

File: src/lidartag.cpp (hunk 1 above)
Comment (knzo25) [reply src/lidartag.cpp]: 削除しました！

File: include/lidartag/lidartag.hpp (hunk 2 above)
Comment (knzo25) [reply include/lidartag/lidartag.hpp:502]: 残念ながらその段階で１−４まだ意味をしません。
しかし、cloud->cloud_edgeにすると少しでもわかりわすくなると思います。
どれでよろしいですか？

File: scripts/rectangle_model.py (hunk 3 above)
Comment (knzo25) [reply scripts/rectangle_model.py]: 削除します！

File: src/rectangle_estimator.cpp (hunk 4 above)
Comment (knzo25) [reply src/rectangle_estimator.cpp]: 削除しました！

**Review by aohsato — APPROVED** [review]

**Pull request merged**